node/manager: synthesize node deletion events

bimmlerd opened the pull request:
When the cilium agent is down (due to a crash or an upgrade), it can miss node events. Upon startup, live nodes are upserted, but when deletions are missed, the agent fails to clean up node-related system state. Examples of such state includes bpf map entries, xfrm states or routes. In particular, the agent fails to clean up node IP to nodeID mappings in the nodeid bpf map. Since K8s will happily recycle such IPs, this can lead to breakage, as the agent associate the wrong nodeID with IPs.

To avoid leaking this state, the node manager now dumps its view of the current set of nodes to a file in the runtime state directory, which can be read on restart of an agent. This is similar to how we restore other state upon restart.

When reading this file, it's important to avoid resurrecting long-gone nodes (as we don't know for how long the agent was down) - instead, we merely take note of which nodes we knew of in the past, compare that to the nodes we consider live (once synced to k8s), and delete the ones which seem to have disappeared.

The motivation to build this reconciliation based on full state dumps to disk is that downstream code generally assumes to have access to a full node object in the deletion callbacks. This makes is infeasible to base the pruning on just the information available in bpf maps. In an alternative design, downstream subsystems are responsible for cleaning up their own state based on just a node identifier, but current code doesn't allow for this.


Fixes: #29822

```release-note
The cilium agent now cleans up stale nodeID mappings and other node-related state on startup
```

bimmlerd committed:
node/manager: synthesize node deletion events

When the cilium agent is down (due to a crash or an upgrade), it can
miss node events. Upon startup, live nodes are upserted, but when
deletions are missed, the agent fails to clean up node-related system
state. Examples of such state includes bpf map entries, xfrm states or
routes. In particular, the agent fails to clean up node IP to nodeID
mappings in the nodeid bpf map. Since K8s will happily recycle such IPs,
this can lead to breakage, as the agent associate the wrong nodeID with
IPs.

To avoid leaking this state, the node manager now dumps its view of the
current set of nodes to a file in the runtime state directory, which can
be read on restart of an agent. This is similar to how we restore other
state upon restart.

When reading this file, it's important to avoid resurrecting long-gone
nodes (as we don't know for how long the agent was down) - instead, we
merely take note of which nodes we knew of in the past, compare that to
the nodes we consider live (once synced to k8s), and delete the ones
which seem to have disappeared.

The motivation to build this reconciliation based on full state dumps to
disk is that downstream code generally assumes to have access to a full
node object in the deletion callbacks. This makes is infeasible to base
the pruning on just the information available in bpf maps. In an
alternative design, downstream subsystems are responsible for cleaning
up their own state based on just a node identifier, but current code
doesn't allow for this.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

gandro commented on pkg/node/manager/manager.go:
```diff
@@ -296,6 +332,10 @@ func (m *manager) Stop(cell.HookContext) error {
 	}
 
+	if m.nodeCheckpointer != nil {
+		m.nodeCheckpointer.Shutdown()
```
Shutting down a trigger doesn't wait for the shutdown function to complete. We probably want to wait on a channel here to block the Stop hook until the snapshot has been performed

bimmlerd replied:
good point, added

gandro commented on pkg/node/manager/manager.go:
```diff
@@ -794,11 +936,24 @@ func (m *manager) NodeDeleted(n nodeTypes.Node) {
-	if !oldNodeExists {
-		m.mutex.Unlock()
-		return
+	// If the node is restored from disk, it doesn't exist in the bookkeeping,
```
Since IPCache is always created from scratch, we don't have to remove any stale entries there. We probably should investigate if we can skip the IPCache logic here and directly go to the `datapath.NodeHandlers` below

bimmlerd replied:
skipping `removeNodeFromIPCache` now - is that enough?

gandro replied:
I see `removeNodeFromIPCache` also calls into the IPSetMgr. I think we already have a reconciler for IPSets which cleans up stale IPSet entries, so skipping `removeNodeFromIPCache` should be fine - but maybe not on stable branches (if we want to backport this). Something to keep in mind.

Other than that, this looks good to me, thanks!

gandro reviewed (commented):
Awesome work! Looks excellent to me. Two very minor things

gandro reviewed (approved):
Looks good to me now!

thorn3r reviewed (approved):
Nice work :+1:

bimmlerd committed:
controlplane: clear environment after test

Clearing the environment in the middle of the test can cause failures
related to state being deleted, as the "environment" being cleared is
simply the StateDir of the agent.

Fixes: 940b186ab4 ("test/controlplane: Fix tests after removal of global hives")

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

nebril commented on pkg/node/manager/manager.go:
```diff
@@ -299,6 +335,15 @@ func (m *manager) Stop(cell.HookContext) error {
+		// for up to its MinInterval, which is too long.
+		m.nodeCheckpointer.Shutdown()
+		close(m.checkpointerDone)
+		_ = m.checkpoint() // Not much we can do on error.
```
Can we at least log the error?

bimmlerd replied:
fair point, added.

nebril reviewed (approved):
LGTM, one minor comment left inline.
